internal/agent/device/console/console_test.go

```diff
@@ -0,0 +1,149 @@
+		mockExecutor = executer.NewMockExecuter(ctrl)
+
+		consoleController = NewConsoleController(mockGrpcClient, deviceName, mockExecutor, logger)
+
+		desired = &api.RenderedDeviceSpec{
+			Console: &api.DeviceConsole{
+				SessionID: sessionId,
+			},
+		}
+
+		testCommand = exec.Command("bash", "-i", "-l")
+	})
+
+	When("Test Console.Sync()", func() {
+		It("no desired console", func() {
+			consoleController.active = true
+
+			err := consoleController.Sync(ctx, &api.RenderedDeviceSpec{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(consoleController.active).To(BeFalse())
+		})
+
+		It("active console with new desired console", func() {
+			consoleController.active = true
+			consoleController.streamClient = mockStreamClient
+
+			err := consoleController.Sync(ctx, desired)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(consoleController.active).To(BeTrue())
+		})
+
+		It("no desired console with active stream failure", func() {
+			consoleController.active = true
+			consoleController.streamClient = mockStreamClient
+
+			mockStreamClient.EXPECT().CloseSend().Return(errors.New("close send error"))
+
+			err := consoleController.Sync(ctx, &api.RenderedDeviceSpec{})
+			Expect(err).To(HaveOccurred())
+			Expect(consoleController.active).To(BeTrue())
+		})
+
+		It("active console with same session ID", func() {
+			consoleController.currentStreamID = sessionId
+
+			mockStreamClient.EXPECT().Recv().Return(nil, nil).AnyTimes()
+			mockStreamClient.EXPECT().Send(gomock.Any()).Return(nil).AnyTimes()
+			mockStreamClient.EXPECT().CloseSend().Return(nil).AnyTimes()
+			mockGrpcClient.EXPECT().Stream(gomock.Any()).Return(mockStreamClient, nil)
+			mockExecutor.EXPECT().CommandContext(gomock.Any(), gomock.Any(), gomock.Any()).Return(testCommand)
+
+			err := consoleController.Sync(ctx, desired)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("console session was closed", func() {
+			consoleController.lastClosedStream = sessionId
+
+			err := consoleController.Sync(ctx, desired)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("no gRPC client available", func() {
+			consoleController.grpcClient = nil
+
+			err := consoleController.Sync(ctx, desired)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("error creating shell process", func() {
+			consoleController.lastClosedStream = ""
+
+			mockGrpcClient.EXPECT().Stream(gomock.Any()).Return(nil, errors.New("shell creation error"))
+			testCommand.Process = nil
+			mockExecutor.EXPECT().CommandContext(gomock.Any(), gomock.Any(), gomock.Any()).Return(testCommand)
+
+			err := consoleController.Sync(ctx, desired)
+			Expect(err).To(HaveOccurred())
+		})
+
+		It("error creating console stream client", func() {
+			mockGrpcClient.EXPECT().Stream(gomock.Any()).Return(nil, errors.New("stream creation error"))
+			mockExecutor.EXPECT().CommandContext(gomock.Any(), gomock.Any(), gomock.Any()).Return(testCommand)
+
+			err := consoleController.Sync(ctx, desired)
+			Expect(err).To(HaveOccurred())
```

Comment:
again it would be good to understand what errored and if that is expected. see how spec manager created explicit error types that can be wrapped and validated. That would make sense here as well.

Reply:
A good example of this can be seen in this search for ErrNoContent. It shows how we can explicitly ensure an error without having to match error strings. This can be a future change just something that will help future tests.

```go
// declare
var (
   ErrNoContent         = fmt.Errorf("no content")
 )
 
 // wrap and return
 if noContent {
     return fmt.Errorf("we failed for some reason: %w")
 }
 
 // test
 require.ErrorIs(err, ErrNoContent)
 ```
ref. https://github.com/search?q=repo%3Aflightctl%2Fflightctl%20ErrNoContent&type=code

Reply:
The provided examples are great, but for these unit tests:

1. The tested logic only distinguish between nil or non-nil errors. The concrete error is irrelevant for the current console implementation.
2. The concrete errors are not duplicated between the test cases.

Even though it's possible to follow your example, I don't see the benefit in these unit tests.